File: utils/glm-client.ts
import axios, { AxiosInstance } from 'axios';
import { getGLMConfig } from '../config/settings';

export class GLMClient {
  private client: AxiosInstance;
  private config: ReturnType<typeof getGLMConfig>;

  constructor() {
    this.config = getGLMConfig();

    this.client = axios.create({
      baseURL: this.config.baseUrl,
      headers: {
        'Authorization': `Bearer ${this.config.apiKey}`,
        'Content-Type': 'application/json',
      },
      timeout: 60000,
    });
  }

  async generate(
    messages: Array<{ role: string; content: string }>,
    stream: boolean = false,
    temperature: number = 0.7,
    maxTokens: number = 4096,
  ): Promise<string> {
    const payload = {
      model: this.config.model,
      messages,
      temperature,
      max_tokens: maxTokens,
      stream,
    };

    try {
      const response = await this.client.post('/chat/completions', payload);

      if (stream) {
        return await this.handleStreamResponse(response.data);
      }

      return response.data.choices[0].message.content;
    } catch (error: any) {
      throw new Error(`GLM API Error: ${error.message}`);
    }
  }

  async generateStream(
    messages: Array<{ role: string; content: string }>,
    temperature: number = 0.7,
    maxTokens: number = 4096,
  ): Promise<AsyncGenerator<string>> {
    const payload = {
      model: this.config.model,
      messages,
      temperature,
      max_tokens: maxTokens,
      stream: true,
    };

    try {
      const response = await this.client.post('/chat/completions', payload, {
        responseType: 'stream',
      });

      return this.parseStream(response.data);
    } catch (error: any) {
      throw new Error(`GLM Stream Error: ${error.message}`);
    }
  }

  private async handleStreamResponse(data: any): Promise<string> {
    let fullContent = '';

    for await (const chunk of data) {
      if (chunk.toString().startsWith('data: ')) {
        const dataStr = chunk.toString().slice(6);
        if (dataStr === '[DONE]') {
          break;
        }

        try {
          const parsed = JSON.parse(dataStr);
          if (parsed.choices && parsed.choices[0]) {
            const delta = parsed.choices[0].delta;
            if (delta.content) {
              fullContent += delta.content;
            }
          }
        } catch {
          continue;
        }
      }
    }

    return fullContent;
  }

  private async *parseStream(data: any): AsyncGenerator<string> {
    for await (const chunk of data) {
      if (chunk.toString().startsWith('data: ')) {
        const dataStr = chunk.toString().slice(6);
        if (dataStr === '[DONE]') {
          break;
        }

        try {
          const parsed = JSON.parse(dataStr);
          if (parsed.choices && parsed.choices[0]) {
            const delta = parsed.choices[0].delta;
            if (delta.content) {
              yield delta.content;
            }
          }
        } catch {
          continue;
        }
      }
    }
  }

  async close(): Promise<void> {
    this.client.defaults.baseURL = '';
  }
}

export interface AsyncGenerator<T> extends AsyncIterableIterator<T> {
  [Symbol.asyncIterator](): AsyncIterator<T>;
}

File: glm_code_system/utils/glm_client.py
"""GLM API client for interacting with GLM models."""


import httpx
from typing import Any, AsyncGenerator

from config.settings import settings


class GLMClient:
    """Client for GLM API interactions."""

    def __init__(
        self,
        api_key: str | None = None,
        model: str | None = None,
        base_url: str | None = None,
    ) -> None:
        """Initialize GLM client."""
        self.api_key = api_key or settings.glm_api_key
        self.model = model or settings.glm_model
        self.base_url = base_url or settings.glm_base_url

        self.client = httpx.AsyncClient(
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json",
            },
            timeout=60.0,
        )

    async def generate(
        self,
        messages: list[dict[str, Any]],
        stream: bool = False,
        temperature: float = 0.7,
        max_tokens: int = 4096,
    ) -> str:
        """Generate response from GLM model."""
        payload = {
            "model": self.model,
            "messages": messages,
            "temperature": temperature,
            "max_tokens": max_tokens,
            "stream": stream,
        }

        response = await self.client.post(
            f"{self.base_url}/chat/completions",
            json=payload,
        )

        response.raise_for_status()
        data = response.json()

        return data["choices"][0]["message"]["content"]

    async def generate_stream(
        self,
        messages: list[dict[str, Any]],
        temperature: float = 0.7,
        max_tokens: int = 4096,
    ) -> AsyncGenerator[str, None]:
        """Generate streaming response from GLM model."""
        payload = {
            "model": self.model,
            "messages": messages,
            "temperature": temperature,
            "max_tokens": max_tokens,
            "stream": True,
        }

        async with self.client.stream(
            "POST",
            f"{self.base_url}/chat/completions",
            json=payload,
        ) as response:
            response.raise_for_status()

            async for line in response.aiter_lines():
                if line.startswith("data: "):
                    data_str = line[6:]
                    if data_str == "[DONE]":
                        break

                    import json

                    try:
                        data = json.loads(data_str)
                        if "choices" in data and len(data["choices"]) > 0:
                            delta = data["choices"][0].get("delta", {})
                            if "content" in delta:
                                yield delta["content"]
                    except json.JSONDecodeError:
                        continue

    async def close(self) -> None:
        """Close the HTTP client."""
        await self.client.aclose()
